src/main/java/org/spongepowered/api/service/user/UserStorage.java

```diff
@@ -0,0 +1,115 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.service.user;
+
+import com.google.common.base.Optional;
+import org.spongepowered.api.GameProfile;
+import org.spongepowered.api.entity.player.Player;
+import org.spongepowered.api.entity.player.User;
+import org.spongepowered.api.service.profile.GameProfileResolver;
+
+import java.util.Collection;
+import java.util.UUID;
+
+/**
+ * Stores the persistent {@link User} data of a {@link Player}.
+ */
+public interface UserStorage {
+
+    /**
+     * Gets the data of a {@link User} by their unique id.
+     *
+     * @param uniqueId The UUID of the user
+     * @return {@link User} or Optional.absent() if not found
+     */
+    Optional<User> get(UUID uniqueId);
+
+    /**
+     * Gets the data of a {@link User} by their last known user name
+     * (case-insensitive).
+     *
+     * <p>To get the current name of a player, use the
+     * {@link GameProfileResolver} service.</p>
+     *
+     * @param uniqueId The UUID of the user
+     * @return {@link User} or Optional.absent() if not found
+     */
+    Optional<User> get(String lastKnownName);
+
+    /**
+     * Gets the data of a {@link User} by their profile.
+     *
+     * @param uniqueId The UUID of the user
+     * @return {@link User} or Optional.absent() if not found
+     */
+    Optional<User> get(GameProfile profile);
+
+    /**
+     * Gets or creates a persistent {@link User} associated with the given
+     * {@link GameProfile}.
+     *
+     * To obtain a {@link GameProfile}, use the {@link GameProfileResolver}.
+     *
+     * @param profile The profile
+     * @return The user object
+     */
+    User getOrCreate(GameProfile profile);
+
+    /**
+     * Deletes the data associated with a {@link User}.
+     *
+     * <p>This might not work if the user is logged in.</p>
+     *
+     * @param profile The profile
+     * @return true if the deletion was successful
+     */
+    boolean delete(GameProfile profile);
+
+    /**
+     * Gets the collection of all {@link GameProfile}s with stored {@link User}
+     * data.
+     *
+     * <p>Note that this method is resource-intensive depending on the amount of
+     * stored data.</p>
+     *
+     * <p>Use {@link #get(GameProfile)} to get the {@link User} data
+     * corresponding to a {@link GameProfile}.</p>
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAll();
```

Comment:
For that use case, we should add similar methods to the profile resolver service:

``getAllCachedProfiles()`` and ``matchCachedProfile(String)``